src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToForEachProvider : AbstractConvertLinqQueryToForEachProvider<QueryExpressionSyntax>
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override string Title => CSharpFeaturesResources.Convert_to_foreach;
+
+        protected override bool TryConvert(
+            QueryExpressionSyntax queryExpression,
+            SemanticModel semanticModel,
+            ISemanticFactsService semanticFacts,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdateInfo)
+                => new Converter(semanticModel, semanticFacts, queryExpression, cancellationToken).TryConvert(out documentUpdateInfo);
+
+        /// <summary>
+        /// Finds a node for the span and checks that it is either a QueryExpressionSyntax or a QueryExpressionSyntax argument within ArgumentSyntax.
+        /// </summary>
+        protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+        {
+            var node = root.FindNode(span);
+            return node as QueryExpressionSyntax ?? (node is ArgumentSyntax argument ? argument.Expression as QueryExpressionSyntax : default);
+        }
+
+        private sealed class Converter
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly ISemanticFactsService _semanticFacts;
+            private readonly CancellationToken _cancellationToken;
+            private readonly QueryExpressionSyntax _source;
+            private readonly List<string> _introducedLocalNames;
+
+            public Converter(SemanticModel semanticModel, ISemanticFactsService semanticFacts, QueryExpressionSyntax source, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _semanticFacts = semanticFacts;
+                _source = source;
+                _introducedLocalNames = new List<string>();
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // Do not try refactoring queries with comments of conditional compilation in them.
```

Comment:
>comments of conditional [](start = 55, length = 23)

Typo "**of**". Sorry it looks like this came from my suggestion. I meant to say "or".